[Core] add auto `device_map` support to pipelines (#6857)
huggingface/diffusers: docs/source/en/training/distributed_inference.md

* it works only on a single GPU
* a single model might not fit on a single GPU `enable_sequential_cpu_offload()` might work but will be extremely slow. Here also we're limited to using a single GPU only.

To use the "balanced" `device_map`, just pass `device_map="balanced"` when instantiating the pipeline:

```diff
from diffusers import DiffusionPipeline
import torch

pipeline = DiffusionPipeline.from_pretrained(
-    "runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16, use_safetensors=True,
+    "runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16, use_safetensors=True, device_map="balanced"
)
image = pipeline("a dog").images[0]
image
```

Pass a dictionary to enforce a restriction on the maximum GPU memory that can be used:

```diff
from diffusers import DiffusionPipeline
import torch

max_memory = {0:"1GB", 1:"1GB"}
pipeline = DiffusionPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5",
    torch_dtype=torch.float16, 
    use_safetensors=True, 
    device_map="balanced",
+    max_memory=max_memory
)
image = pipeline("a dog").images[0]
image

> ```

Reviewer: Maybe mention by default that it is going the use the maximum memory of all devices (e.g. all gpus). Also it might be good to mention that if the models doesn't fit on the gpus, we will offload the model to the cpu. If they don't have enough space on the cpu, we might have an error ;)

</Tip>

> ### Using `device_map`

Reviewer 2: I think this'll be clearer for users.

```suggestion
### Device placement
```

> <Tip warning={true}>
> 
> 🧪 This feature is experimental and its APIs might change in future. 
> 
> </Tip>

Reviewer 2: Let's start using this format more since it takes up less space and is cleaner :)

```suggestion
> [!WARNING]
> This feature is experimental and its APIs might change in the future. 
```

> With 🤗 Accelerate, you can make use of a supported `device_map` strategy to distribute the model-level components of a pipeline across multiple devices. The section below explains a situation where performing distributed inference via `device_map` could be useful.

Reviewer 2: ```suggestion
With Accelerate, you can use the `device_map` to determine how to distribute the model-level components of a pipeline across multiple devices. This is useful in situations where you have more than one GPU.
```

Author: `model-level components` is a preference as I think it's clearer.

> Let's say you have two consumer GPUs each having 8GBs of VRAM. Using `enable_model_cpu_offload()` might not work best here because -
> 
> * it works only on a single GPU
> * a single model might not fit on a single GPU `enable_sequential_cpu_offload()` might work but will be extremely slow. Here also we're limited to using a single GPU only.

Reviewer 2: ```suggestion
For example, if you have two 8GB GPUs, then using [`~DiffusionPipeline.enable_model_cpu_offload`] may not work so well because:

* it only works on a single GPU
* a single model might not fit on a single GPU ([`~DiffusionPipeline.enable_sequential_cpu_offload`] might work but it will be extremely slow and it is also limited to a single GPU)
```

Reviewer 3: If I understand this PR correctly, each model is completely assigned to a GPU, or to the CPU as a whole. How does this solve the problem of a single model not fitting on a single GPU?

Author: If it doesn't fit on a single GPU it's offloaded to the CPU, yes. If it doesn't fit on the CPU, yes that is going to be a problem for now. We will error out in that case (have noted that in the docs). This is gonna be improved in the future iterations too (disk offloading) if there's enough demand. 

If you're thinking about model sharding across GPU, CPU, etc. then it's a separate PR: https://github.com/huggingface/diffusers/pull/6396

> To use the "balanced" `device_map`, just pass `device_map="balanced"` when instantiating the pipeline:

Reviewer 2: Let's add a little info about what the "balanced" strategy is

```suggestion
To make use of both GPUs, you can use the "balanced" device placement strategy which evenly splits the model across all available GPUs.

> [TIP]
> Only the "balanced" strategy is supported at the moment, and we plan to support additional mapping strategies in the future.
```

```diff
from diffusers import DiffusionPipeline
import torch

pipeline = DiffusionPipeline.from_pretrained(
-    "runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16, use_safetensors=True,
+    "runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16, use_safetensors=True, device_map="balanced"
)
image = pipeline("a dog").images[0]
image
```

> Pass a dictionary to enforce a restriction on the maximum GPU memory that can be used:

Reviewer 2: ```suggestion
You can also pass a dictionary to enforce the maximum GPU memory that can be used on each device:
```

```diff
from diffusers import DiffusionPipeline
import torch

max_memory = {0:"1GB", 1:"1GB"}
pipeline = DiffusionPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5",
    torch_dtype=torch.float16, 
    use_safetensors=True, 
    device_map="balanced",
+   max_memory=max_memory
)
image = pipeline("a dog").images[0]
image
```

> By default, Diffusers will use the maximum memory of all devices (i.e., all GPUs). If the models don't fit on the GPUs, they will be offloaded to the CPU. If the CPU doesn't have enough memory, you might see an error.

Reviewer 2: I think it'd be nice to suggest some way to handle the error instead of leaving it hanging as is

```suggestion
By default, Diffusers uses the maximum memory of all devices. If the models don't fit on the GPUs, they are offloaded to the CPU. If the CPU doesn't have enough memory, then you might see an error.
```

> You can reset any `device_map` of a pipeline by calling `reset_device_map()` on the pipeline object.
> 
> To use methods like `to()`, `enable_sequential_cpu_offload()`, and `enable_model_cpu_offload()` on a supported pipeline that has been device-mapped, first call `reset_device_map()` on it and then use these methods.

Reviewer 2: ```suggestion
Call `reset_device_map` to reset the `device_map` of a pipeline. This is also necessary if you want to use methods like `to()`, [`~DiffusionPipeline.enable_sequential_cpu_offload`], and [`~DiffusionPipeline.enable_model_cpu_offload`] on a pipeline that was device-mapped.

```py
pipeline.reset_device_map()
```
```

> To use methods like `to()`, `enable_sequential_cpu_offload()`, and `enable_model_cpu_offload()` on a supported pipeline that has been device-mapped, first call `reset_device_map()` on it and then use these methods.

Author: Cc: @yiyixuxu here. Might be relevant as well: https://github.com/huggingface/diffusers/pull/6857/#discussion_r1545895124

> > [!NOTE]  
> > Currently, we support only "balanced" `device_map`. We plan to support more device mapping strategies in future.

Reviewer 2: This format is only supported for `TIP` and `WARNING` in the docs. I think we should also move this up to where we first introduce the "balanced" strategy.

```suggestion
```

</Tip>

### Device placement

> [!WARNING]
> This feature is experimental and its APIs might change in the future. 

With Accelerate, you can use the `device_map` to determine how to distribute the models of a pipeline across multiple devices. This is useful in situations where you have more than one GPU.

For example, if you have two 8GB GPUs, then using [`~DiffusionPipeline.enable_model_cpu_offload`] may not work so well because:

* it only works on a single GPU

> * a single model might not fit on a single GPU ([`~DiffusionPipeline.enable_sequential_cpu_offload`] might work but it will be extremely slow and it is also limited to a single GPU)

Reviewer 3: I think I made a comment yesterday about this but can't find it now. How does this PR help when a model does not fit in a single GPU?

Author: Yeah, that's a GitHub bug, apparently. I responded to it but it got disappeared as well. 

We offload to the CPU. In case there's not enough CPU memory, we error out. 

The next alternative would be to split the model across devices - GPU, CPU, and disk. This is a separate PR: https://github.com/huggingface/diffusers/pull/6396.
